packages/discount-code-exporter/src/main.js

```diff
@@ -1,3 +1,158 @@
+  accessToken: string;
+  delimiter: string
+  exportFormat: string;
+  predicate: string;
+  multiValueDelimiter: string
+  _processCode: Function;
+
+  constructor (
+    options: ImporterOptions,
+    logger: LoggerOptions,
+  ) {
+    if (!options.apiConfig)
+      throw new Error('The contructor must be passed an `apiConfig` object')
+    if (options.batchSize > 500)
+      throw new Error('The `batchSize` must not be more than 500')
+    this.apiConfig = options.apiConfig
+    this.client = createClient({
+      middlewares: [
+        createAuthMiddlewareForClientCredentialsFlow(this.apiConfig),
+        createUserAgentMiddleware({
+          libraryName: pkg.name,
+          libraryVersion: pkg.version,
+        }),
+        createHttpMiddleware({ host: this.apiConfig.apiUrl }),
+      ],
+    })
+
+    const defaultOptions = {
+      accessToken: '',
+      batchSize: 500,
+      delimiter: ',',
+      exportFormat: 'json',
+      multiValueDelimiter: ';',
+      predicate: '',
+    }
+    Object.assign(this, defaultOptions, options)
+
+    this.logger = {
+      error: () => {},
+      warn: () => {},
+      info: () => {},
+      verbose: () => {},
+      ...logger,
+    }
+
+    this._processCode = this._processCode.bind(this)
+  }
+
+  run (outputStream: stream$Writable) {
+    this.logger.info('Starting Export')
+    if (this.exportFormat === 'csv') {
+      const csvOptions = {
+        headers: true,
+        delimiter: this.delimiter,
+      }
+      const csvStream = csv
+        .createWriteStream(csvOptions)
+        .transform(this._processCode)
+      csvStream.pipe(outputStream)
+      this._fetchCodes(csvStream)
+      .then(() => {
+        this.logger.info('Operation completed successfully')
+        csvStream.end()
+      })
+      .catch((e: Error) => {
+        outputStream.emit('error', e)
+      })
+    } else {
+      const jsonStream = JSONStream.stringify()
+      jsonStream.pipe(outputStream)
+
+      this._fetchCodes(jsonStream)
+      .then(() => {
+        this.logger.info('Operation completed successfully')
+        if (outputStream !== process.stdout)
+          jsonStream.end()
+      })
+      .catch((e: Error) => {
+        outputStream.emit('error', e)
+      })
+    }
+  }
+
+  _fetchCodes (output: stream$Writable) {
+    const request = this._buildRequest()
+    return this.client.process(request, (data) => {
+      if (data.statusCode !== 200)
+        return Promise.reject(data)
+      this.logger.verbose(`Successfully exported ${data.body.count} codes`)
+      data.body.results.forEach((codeObj) => {
+        output.write(codeObj)
+      })
+      return Promise.resolve()
+    }, { accumulate: false })
+  }
+
+  _buildRequest () {
+    const service = this._createService()
+    const uri = service.perPage(this.batchSize)
+    if (this.predicate)
+      uri.where(this.predicate)
+    const request: Object = {
+      uri: uri.build(),
+      method: 'GET',
+    }
+    if (this.accessToken)
```

Comment:
Doesn't it make sense to _require_ `accessToken` already on instantiation?
then this if-statement would no longer be needed